### PR TITLE
Added format-x support for ipc_module, tested with demo config

### LINKED ISSUE #2775: Native styling support for custom IPC modules
### Discussed in https://github.com/polybar/polybar/discussions/2771

<div type='discussions-op-text'>

<sup>Originally posted by **beyond9thousand** July 21, 2022</sup>
Would love to be able to assign different colors to different hooks in a custom IPC module</div>

---

That could probably be done by supporting `format-X` for every `hook-X` and have `format-X` inherit from `format`.

PRs welcome :)

### PR DESCRIPTION
**Pull request by madhavpcm**
## What type of PR is this? (check all applicable)

* [ ] Refactor
* [x] Feature
* [ ] Bug Fix
* [ ] Optimization
* [x] Documentation Update
* [ ] Other: *Replace this with a description of the type of this PR*

## Description

Users can now assign separate formats for different hooks inside the IPC module.

## Related Issues & Documents

Closes #2775 
## Documentation (check all applicable)

* [x] This PR requires changes to the Wiki documentation (describe the changes)
* [ ] This PR requires changes to the documentation inside the git repo (please add them to the PR).
* [ ] Does not require documentation changes

Changes to documentation
-----
In wiki > Modules > ipc
Examples,
```
[module/demo]
type = custom/ipc
hook-0 = echo foobar
hook-1 = date +%s
hook-2 = whoami
format= <label>
format-0-foreground= ${color.foreground}
format-1-foreground= ${color.foreground}
format-2-foreground= ${color.foreground}
; Based on colors defined by user
format-0-background= ${color.shade1}
format-1-background= ${color.shade2}
format-2-background= ${color.shade3}
initial = 1
click-left = "#demo.hook.0"
click-right = "#demo.hook.1"
double-click-left = "#demo.hook.2"

```

### CODE REVIEW

File: src/modules/ipc.cpp
Comment (patrick96) [inline src/modules/ipc.cpp:117]: This crashes the module if `initial` is not set, or set to 0.

We guarantee that `-1 <= m_current_hook < m_hooks.size()`. So, we can simply return `DEFAULT_FORMAT` if `has_hook()` returns false and otherwise return the constructed format name.

File: src/modules/ipc.cpp
Comment (patrick96) [inline src/modules/ipc.cpp:67]:
```suggestion
```

**Review by patrick96 — CHANGES_REQUESTED** [review]
Thanks :)

I have unfortunately noticed that this breaks any existing config that does formatting:

```dosini
[module/ipc]
type = custom/ipc

hook-0 = echo foobar

format-background = #f00
format-foreground = #fff
```

The current behavior for this is that the module will display `foobar` with white text and red background.
However, with this PR, those settings are ignored (since we are now using `format-0`) and the module displays `foobar` with the bar background and foreground colors.

This makes things more difficult than I first thought. At the very least, we should fall back to `format` if `format-N` is not defined.

---

Please also add a changelog entry for your changes: https://github.com/polybar/polybar/blob/master/CONTRIBUTING.md#changelog

And run `clang-format` on the files you modified: https://polybar.readthedocs.io/en/latest/dev/style-guide.html#code-formatting

**Comment by madhavpcm** [comment]
Working on it asap.

File: src/modules/ipc.cpp
Comment (madhavpcm) [reply src/modules/ipc.cpp:117]: I think we need to keep some sort of mapping (like unordered_map) for hook-x to the internal index we store.

Comment (patrick96) [reply src/modules/ipc.cpp:117]: A simple fallback could also be implemented by adding the `format-N` using `m_formatter->add_optional` and then in `get_format` first checking `m_formatter->has_format` and if it doesn't exist, fall back to the default format.

**Comment by patrick96** [comment]
> Working on it asap.

There is really no need for stressing about completing this. Take your time, we're not in a hurry :)

File: src/modules/ipc.cpp
Comment (madhavpcm) [reply src/modules/ipc.cpp:117]: > A simple fallback could also be implemented by adding the `format-N` using `m_formatter->add_optional` and then in `get_format` first checking `m_formatter->has_format` and if it doesn't exist, fall back to the default format.

I tried this, add_optional is not adding the format to the list, in addition , I noticed that this config
```
hook-0 = echo foobar
hook-1 = date +%s
;hook-2 = whoami
hook-3 = pwd
format= <label>
format-0-foreground= ${color.foreground}
format-1-foreground= ${color.foreground}
;format-2-foreground= ${color.foreground}

format-0-background= ${color.shade1}
format-1-background= ${color.shade2}
;format-2-background= ${color.shade3}
initial = 1
click-left = "#demo.hook.0"
click-right = "#demo.hook.1"
;double-click-left = "#demo.hook.2"
```
doesnt detect hook-3, because hook-2 stuff is commented. ( is this a rule in configs where you define hook-x in increasing order from 0 to n).  Im trying to figure out why m_conf.has(m_modname, name) is failing.

Comment (patrick96) [reply src/modules/ipc.cpp:117]: > I tried this, add_optional is not adding the format to the list,

Yeah, the disadvantage of that is that it only detects a format if the key `format-N` is defined. If you only define `format-N-foreground`, it doesn't see it.

And yes, it's a rule that lists like this have to be complete. So for `hook-3` to be loaded, `hook-0` up to `hook-2` have to exist as well.

Comment (madhavpcm) [reply src/modules/ipc.cpp:117]: > Yeah, the disadvantage of that is that it only detects a format if the key `format-N` is defined. If you only define `format-N-foreground`, it doesn't see it.

If I add format-N in the config, there is no output from the IPC module. In this case, neither ```add()``` nor ```add_optional()``` works. If I search for format-n in  ipc::constructor with ```has_format ("format-n")```, the result turns out negative.

Comment (patrick96) [reply src/modules/ipc.cpp:117]: Then there is probably something else going wrong. If the setting `format-0` exists, both `add("format-0")` and `add_optional("format-0")` will load the format and `has_format("format-0")` will return true.

Comment (madhavpcm) [reply src/modules/ipc.cpp:117]: > Then there is probably something else going wrong. If the setting `format-0` exists, both `add("format-0")` and `add_optional("format-0")` will load the format and `has_format("format-0")` will return true.

polybar|error: Failed to get contents for "module/demo" (err: Format "format-0" has not been added)
, throw undefined_format is being thrown. This is what happens when I define format-0 in config, ```has_format()``` fails, 
image.

Comment (patrick96) [reply src/modules/ipc.cpp:117]: How are you adding the formats to the formatter?

> My doubt is, why does m_formats only have "format", shouldnt it have the other keywords of module/demo

`m_formats` only contains the formats added with `add` or `add_optional`.

`m_conf` is the thing that contains all config options

Comment (madhavpcm) [reply src/modules/ipc.cpp:117]: > How are you adding the formats to the formatter?
In this case, I commented any addition of format. But even if I add, with either add or add_optional (after defining format-n in config), the module is not visible. I guess this error was due to the fact that I wasn't adding the format.

**Comment by patrick96** [comment]
I think configuration 2 above is the way to go here. There is no good way to make it work without defininig `format-X` since its difficult to tell whether any other format properties are defined.

As to why the module doesn't show anything for that, I don't know. If you upload the latest changes, I may be able to take a look.

> Works with colors of the main bar, hence what I observed is that we should only call add() if there exists format-x in the config.
For this I have to traverse elements in m_conf->m_sections(m_name) and check if format-x is there, as of now I didnt find any function which does this

The thing you're looking for is `m_conf.has`. But what you are describing is exactly what `add_optional` does: it checks if the format exists and only adds it if it does.

**Comment by madhavpcm** [comment]
> As to why the module doesn't show anything for that, I don't know. If you upload the latest changes, I may be able to take a look.

Latest changes demonstrate that if format-X is defined, the module just doesnt output anything (as if it crashed). Comment format-0, so it will reach hook-0 output, from there if you trigger any other hook it hides itself.

**Comment by madhavpcm** [comment]
> > As to why the module doesn't show anything for that, I don't know. If you upload the latest changes, I may be able to take a look.
> 
> Latest changes demonstrate that if format-X is defined, the module just doesnt output anything (as if it crashed). Comment format-0, so it will reach hook-0 output, from there if you trigger any other hook it hides itself.
However if you change add_optional to add, and then place this in get format instead
 ```cpp  
  if (m_conf.has(m_name, format_i)) {
        return DEFAULT_FORMAT;
      } else {
        return format_i;
      }
```
I notice that, if format is defined, or format-x is defined, and get_format returns the defined format in interest, then the module fails.

**Comment by patrick96** [comment]
Ah, I see what the issue is. `<label>` doesn't exist in the ipc module. Try using `format-X = <output>`

File: src/modules/ipc.cpp
Comment (patrick96) [inline src/modules/ipc.cpp:117]:
```suggestion
      if (m_formatter->has_format(format_i)) {
        return format_i;
      }
```

Comment (madhavpcm) [reply src/modules/ipc.cpp:117]: Sure, il follow up , thanks again!

**Comment by madhavpcm** [comment]
```
format= <output>
format-background = #ff2
format-0 = <output>
format-1 = <output>
format-2 = <output>

format-0-foreground= ${color.foreground}
format-1-foreground= #f00
format-2-foreground= ${color.foreground}
; Based on colors defined by user
format-0-background= ${color.shade1}
format-2-background= ${color.shade3}
```
in this particular config,I want say the partial config of format defined here and partial config of a format-x, to be used in the output. However  format-1-background is not defined, its falling back to bars default color. Is it possible to mix things up ? I tried look around the render function calls , didnt notice at first glance.

**Comment by madhavpcm** [comment]
Also, since <output> is deprecated in some other modules, I can also work on <label> support for IPC module.

**Comment by patrick96** [comment]
> Is it possible to mix things up ?

No. When any property on a format is not defined, it falls back to how that property is defined in the [`[settings]` section](https://github.com/polybar/polybar/wiki/Configuration#application-settings). I think at some point we'll want to have some better fallback mechanisms for formats (there is already an issue about that: #2022), but that would be a bigger undertaking.

> Also, since output is deprecated in some other modules, I can also work on label support for IPC module.

That would be cool. I think the ipc module is the only one that doesn't have any labels. But wait until we merged this PR.

**Review by patrick96 — CHANGES_REQUESTED** [review]
Alright, this seems to be working now :smiley: 

Plase also add a changelog entry and then we can merge this. https://github.com/polybar/polybar/blob/master/CONTRIBUTING.md#changelog

**Review by patrick96 — APPROVED** [review]

**Pull request merged**
